### DISCUSSION TITLE
Making a GraphQL request before building a dynamic page

### DISCUSSION THREAD

**Discussion by tiffanynk** [vercel/next.js#41260] [Help]
Hello,

I'm fairly new to Next.js and I'm having some trouble figuring out a solution for our product. We have an application that is used by different companies, and the users within those companies are able to access various content depending on their permissions. We're adding a product offering so those companies can choose between two default themes (light mode / dark mode) as a default. Individual users can still manually update their preferences. Both the company preference and user's are saved in the DB.

Right now, we only have the user's preference available to toggle. We hacked our way into a solution to prevent a flash of unstyled content (where the screen flashes white for light mode then goes to dark mode) by saving the value in `localStorage` and getting it in a script in `_document.tsx`:

```
render( ) {
   const setIfLightMode = {... 
         get value from local storage and set a className on <body> for lightMode if it exists
    ...}
...
   return (
      ......
      <body>
          <script dangerouslySetInnerHTML={{ __html: setIfLightMode }} />
          <Main />
          <NextScript />
      </body>
```

I have a resolver written in the backend to handle the conflict between a company preference and user preference, but I'm not sure how to handle setting this className by using a GQL hook i.e. `useGetThemePreferenceQuery` before build, to prevent the flash. Has anyone handled any similar issues?

**Comment by gaston-flores** [comment]
You can't use a hook for this. To have this value available during the intial render, you'll need to make it in [`getServerSideProps`/`getInitialProps`](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props) using a GQL client, and pass `setLightMode` as a prop to your page.
